### PR TITLE
Abdulhafiz96/implement post signup & signin

### PR DESCRIPTION
**Pull request by abdulhafiz96**

### CODE REVIEW

**Review by dilarafirtina — APPROVED** [review]

**Review by dilarafirtina — APPROVED** [review]

File: src/controllers/auth.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    (await Volunteer.findOne({ email })) || (await Ngo.findOne({ email }));
+  if (!user) {
+    return next(new ErrorResponse('User not found', 404));
+  }
+  const isMatch = await bcrypt.compare(password, user.password);
+  if (!isMatch) {
+    return next(new ErrorResponse('Invalid credentials', 401));
+  }
+  const token = jwt.sign(
+    { email: user.email, username: user.username },
```
Comment (muslimomar) [inline src/controllers/auth.js:53]: For the payload, let's store `user._id` only.

```javascript
       { _id: user._id.toHexString() };
```

File: src/controllers/auth.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    process.env.JWT_SECRET,
+    {
+      expiresIn: '14h',
```
Comment (muslimomar) [inline src/controllers/auth.js:56]: make the expiration 14 days

File: src/controllers/auth.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+  );
+  res.cookie('token', token, {
+    expires: new Date(Date.now() + 10 * 24 * 60 * 60 * 1000),
```
Comment (muslimomar) [inline src/controllers/auth.js:60]: make the expiration 14 days

File: src/middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const jwt = require('jsonwebtoken');
+
+const isAuth = (req, res, next) => {
+  if (req.cookies) {
+    const { token } = req.cookies;
+    if (!token) {
+      res.sendStatus(401);
+    } else {
+      req.user = jwt.verify(token, process.env.JWT_SECRET);
```
Comment (muslimomar) [inline src/middlewares/auth.js:9]: surround the `jwt.verify` with a `try/catch`, and in `catch` block, send a message to the user, about an invalid/expired token.
something like this: 

```javascript
       try {
        req.user = jwt.verify(token, process.env.JWT_SECRET);
      } catch (err) {
        return next(new ErrorResponse('Invalid/expired token', 401));
      }
```

File: src/middlewares/express-validator-err.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+const { validationResult } = require('express-validator');
+
+exports.validate = (req, res, next) => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return res.status(400).json({ errors: errors.array() });
+  }
+  return next();
+};
```
Comment (muslimomar) [inline src/middlewares/express-validator-err.js:9]: amazing job utilizing a validator middleware 💪

File: src/middlewares/express-validator-err.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+const { validationResult } = require('express-validator');
+
```
Comment (muslimomar) [inline src/middlewares/express-validator-err.js:2]: let's give this file another name, like `bodyValidator.js`

File: src/middlewares/validatorSchemas.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    .withMessage('Password must be at least 8 characters long'),
+  body('password2')
+    .optional()
+    .custom((value, { req }) => {
+      if (value !== req.body.password) {
+        throw new Error('Passwords do not match');
+      }
+      return true;
+    })
+    .withMessage('passwords do not match'),
```
Comment (muslimomar) [inline src/middlewares/validatorSchemas.js:26]: `password2` could be removed.

File: src/routes/auth.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+const ErrorResponse = require('../utils/errorResponse');
+const { validate } = require('../middlewares/express-validator-err');
+const authController = require('../controllers/auth');
+const {
+  validateSignup,
+  validateSignin,
+} = require('../middlewares/validatorSchemas');
+
+router.post('/signup/:type', validateSignup, validate, authController.signup);
+router.post('/login', validateSignin, validate, authController.login);
```
Comment (muslimomar) [inline src/routes/auth.js:11]: Splendid work again 👏

**Review by muslimomar — CHANGES_REQUESTED** [review]
The efforts you've put into this issue are commendable and truly appreciated. I'm glad I have you on my team 👏💪

**Comment by abdulhafiz96** [comment]
thank you so much for the amazing feedback, I have fixed all the issues you mentioned

**Review by muslimomar — APPROVED** [review]
Outstanding work! 🔥

**Pull request merged**